vpn/tunnel.go

```diff
@@ -129,6 +147,61 @@ func (t *Tunnel) ApplyNetworkSettings(ctx context.Context, ns *NetworkSettingsRe
 	return nil
 }
 
+func (t *Tunnel) Update(update *proto.WorkspaceUpdate) error {
+	msg := &TunnelMessage{
+		Msg: &TunnelMessage_PeerUpdate{
+			PeerUpdate: convertWorkspaceUpdate(update),
+		},
+	}
+	select {
+	case <-t.ctx.Done():
+		return t.ctx.Err()
+	case t.sendCh <- msg:
+	}
+	return nil
+}
+
+func (t *Tunnel) start(req *StartRequest) error {
+	rawURL := req.GetCoderUrl()
+	if rawURL == "" {
+		return xerrors.New("missing coder url")
+	}
+	svrURL, err := url.Parse(rawURL)
+	if err != nil {
+		return xerrors.Errorf("parse url: %w", err)
+	}
+	apiToken := req.GetApiToken()
+	if apiToken == "" {
+		return xerrors.New("missing api token")
+	}
+	var header http.Header
+	for _, h := range req.GetHeaders() {
+		header.Add(h.GetName(), h.GetValue())
+	}
+
+	t.conn, err = t.client.NewConn(
```

Comment:
probably worth checking `t.conn` is nil here, just in case the client has a protocol error and sends us multiple start calls.